pkg/config/loader.go

```diff
@@ -10,33 +10,64 @@ import (
 
 	"github.com/go-viper/mapstructure/v2"
 	"github.com/mitchellh/go-homedir"
+	"github.com/spf13/pflag"
 	"github.com/spf13/viper"
 
 	"github.com/golangci/golangci-lint/pkg/exitcodes"
 	"github.com/golangci/golangci-lint/pkg/fsutils"
 	"github.com/golangci/golangci-lint/pkg/logutils"
 )
 
-type FileReader struct {
-	log            logutils.Log
-	cfg            *Config
-	commandLineCfg *Config
+var errConfigDisabled = errors.New("config is disabled by --no-config")
+
+type LoaderOptions struct {
+	Config   string // Flag only. The path to the golangci config file, as specified with the --config argument.
+	NoConfig bool   // Flag only.
+}
+
+type Loader struct {
+	opts LoaderOptions
+
+	viper *viper.Viper
+	fs    *pflag.FlagSet
+
+	log logutils.Log
+
+	cfg *Config
 }
 
-func NewFileReader(toCfg, commandLineCfg *Config, log logutils.Log) *FileReader {
-	return &FileReader{
-		log:            log,
-		cfg:            toCfg,
-		commandLineCfg: commandLineCfg,
+func NewLoader(log logutils.Log, v *viper.Viper, fs *pflag.FlagSet, opts LoaderOptions, cfg *Config) *Loader {
+	return &Loader{
+		opts:  opts,
+		viper: v,
+		fs:    fs,
+		log:   log,
+		cfg:   cfg,
 	}
 }
 
-func (r *FileReader) Read() error {
-	// XXX: hack with double parsing for 2 purposes:
-	// 1. to access "config" option here.
-	// 2. to give config less priority than command line.
+func (r *Loader) Load() error {
```

Comment:
Should we update the receiver variable since `Reader` got renamed to `Loader`?

```suggestion
func (l *Loader) Load() error {
```

(Probably benefits from LSP capabilities to do proper renaming, I won't suggest changes in more places)